langserver/workspace_refs.go

```diff
@@ -0,0 +1,173 @@
+			if err != nil {
+				log.Printf("externalRefsFromPkg: %v: %v\n", pkg, err)
+			}
+		}(pkg)
+	}
+	run.Wait()
+
+	sort.Sort(&results) // sort to provide consistent results
+
+	// TODO: We calculate all the results and then throw them away. If we ever
+	// decide to begin using limiting, we can improve the performance of this
+	// dramatically. For now, it lives in the spec just so that other
+	// implementations are aware it may need to be done and to design with that
+	// in mind.
+	if len(results.results) > params.Limit && params.Limit > 0 {
+		results.results = results.results[:params.Limit]
+	}
+	return results.results, nil
+}
+
+// externalRefsFromPkg collects all the external references from the specified
+// package and returns the results.
+func (h *LangHandler) externalRefsFromPkg(ctx context.Context, bctx *build.Context, conn JSONRPC2Conn, pkg string, rootPath string, results *refResultSorter) (err error) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "externalRefsFromPkg")
+	defer func() {
+		if err != nil {
+			ext.Error.Set(span, true)
+			span.SetTag("err", err.Error())
+		}
+		span.Finish()
+	}()
+	span.SetTag("pkg", pkg)
+
+	// Import the package.
+	buildPkg, err := bctx.Import(pkg, rootPath, 0)
+	if err != nil {
+		if !(strings.Contains(err.Error(), "no buildable Go source files") || strings.Contains(err.Error(), "found packages") || strings.HasPrefix(pkg, "github.com/golang/go/test/")) {
+			return fmt.Errorf("skipping possible package %s: %s\n", pkg, err)
+		}
+		return nil
+	}
+
+	pkgInWorkspace := func(path string) bool {
+		return PathHasPrefix(path, h.init.RootImportPath)
+	}
+
+	// Perform type checking.
+	fs, prog, diags, err := h.cachedTypecheck(ctx, bctx, buildPkg)
+	if err != nil {
+		return err
+	}
+	if len(diags) > 0 {
+		if err := h.publishDiagnostics(ctx, conn, diags); err != nil {
+			return fmt.Errorf("sending diagnostics: %s", err)
+		}
+	}
+
+	// Compute external references.
+	cfg := &refs.Config{
+		FileSet:  fs,
+		Pkg:      prog.Package(pkg).Pkg,
+		PkgFiles: prog.Package(pkg).Files,
+		Info:     &prog.Package(pkg).Info,
+	}
+	refsErr := cfg.Refs(func(r *refs.Ref) {
+		// If the reference itself is located in a vendor folder, exclude it.
+		locFilename := r.Position.Filename
+		if strings.HasPrefix(locFilename, "vendor/") || strings.Contains(locFilename, "/vendor/") {
+			return
+		}
+
+		var defName, defContainerName string
+		if fields := strings.Fields(r.Def.Path); len(fields) > 0 {
+			defName = fields[0]
+			defContainerName = strings.Join(fields[1:], " ")
+		}
+
+		defPkg, err := bctx.Import(r.Def.ImportPath, rootPath, build.FindOnly)
+		if err != nil {
+			// Log the error, and flag it as one in the trace -- but do not
+			// halt execution (hopefully, it is limited to a small subset of
+			// the data).
+			ext.Error.Set(span, true)
+			err := fmt.Errorf("externalRefsFromPkg: failed to import %v: %v", r.Def.ImportPath, err)
```

Comment:
Nope, because the dirs we're trying to import are always Go packages. That said, I did find that it spews a few unactionable errors _for cgo packages_ elsewhere. But I think we do not have a good story for CGO yet, so I'm going to leave as-is for now.:

```
$ echo '{"limit":1}' | LSP_PROXY_TRACE_FS_REQUESTS=true LSP_PROXY=':dev:' src xlang send 'git://github.com/kubernetes/kubernetes?master' 'go' 'workspace/reference'

# Fetching code from codeload.github.com, not gitserver
# using in-process LSP proxy
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/auth: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/auth.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/client: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/client.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/configmap: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/configmap.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/garbagecollector: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/garbagecollector.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/kubectl: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/kubectl.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/master: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/master.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/persistentvolumes: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/persistentvolumes.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/pods: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/pods.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/quota: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/quota.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/replicaset: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/replicaset.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/replicationcontroller: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/replicationcontroller.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/scheduler: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/scheduler.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/secrets: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/secrets.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/serviceaccount: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/serviceaccount.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/storageclasses: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/storageclasses.
Warning: fetching deps for dir /src/k8s.io/kubernetes/test/integration/thirdparty: no buildable Go source files in /src/k8s.io/kubernetes/test/integration/thirdparty.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/Azure/go-ansiterm/winterm: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/Azure/go-ansiterm/winterm.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/coreos/pkg/dlopen: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/pkg/dlopen.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/google/cadvisor/api: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/google/cadvisor/container/docker: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/google/cadvisor/container/libcontainer: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/google/cadvisor/container/raw: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/google/cadvisor/container/rkt: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/google/cadvisor/http: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/google/cadvisor/manager: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/google/cadvisor/manager/watcher/raw: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/google/cadvisor/manager/watcher/rkt: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/google/cadvisor/pages: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/google/cadvisor/validate: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/opencontainers/runc/libcontainer: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/coreos/go-systemd/util.
Warning: fetching deps for dir /src/k8s.io/kubernetes/vendor/github.com/seccomp/libseccomp-golang: no buildable Go source files in /src/k8s.io/kubernetes/vendor/github.com/seccomp/libseccomp-golang.
unexpected EOF
```